expr_checker.py

```diff
@@ -0,0 +1,128 @@
+            return ast.Call(
+                func=ast.Name(self.getattr, ast.Load()),
+                args=[node.func.value, ast.Constant(node.func.attr), node],
+                keywords=[]
+            )
+
+        else:
+            self.generic_visit(node.func)
+
+        return ast.Call(
+            func=ast.Name(self.check_fn, ctx=ast.Load()),
+            args=[node.func, ast.Name(
+                self.check_type_fn, ctx=ast.Load())] + node.args,
+            keywords=node.keywords
+        )
+
+    def visit_FunctionDef(self, node):
+        self.generic_visit(node)
+        if node.name in self.reserved_name:
+            raise NameError(f"safe_eval: {node.name} is a reserved name")
+
+        return node
+
+    def visit_Attribute(self, node):
+        self.generic_visit(node.value)
+
+        if isinstance(node.ctx, ast.Load):
+            subcall = ast.Call(
+                func=ast.Name(self.getattr, ctx=ast.Load()),
+                args=[node.value, ast.Constant(node.attr), node],
+                keywords=[]
+            )
+
+            return ast.Call(
+                func=ast.Name(self.check_type_fn, ctx=ast.Load()),
+                args=[ast.Constant("attribute"), subcall],
+                keywords=[]
+            )
+
+        elif isinstance(node.ctx, ast.Store):
+            raise ValueError("safe_eval: doesn't permit you to store values in attributes")
+
+        elif isinstance(node.ctx, ast.Del):
+            raise ValueError("safe_eval: doesn't permit you to delete attributes")
+
+    def visit_Assign(self, node):
+        self.generic_visit(node.value)
+
+        for target in node.targets:
+            if isinstance(target, ast.Name):
+                if target.id in self.reserved_name:
+                    raise NameError(f"safe_eval: {target.id} is a reserved name")
+            elif isinstance(target, ast.Attribute):
+                raise ValueError("safe_eval: doesn't permit you to modify an attribute")  # FIXME with ast_set_attr
+            elif isinstance(target, ast.Subscript) and isinstance(target.value,
+                                                                  ast.Call) and target.value.func.id in ["globals",
+                                                                                                         "locals"]:
+                raise ValueError("safe_eval: doesn't permit you to modify locals() and globals()")
+            else:
+                raise NotImplementedError(f"{ast.dump(target, indent=4)}")
+
+        return node
+
+    def visit_Subscript(self, node):
+        self.generic_visit(node)
+        return ast.Call(
+            func=ast.Name(self.check_type_fn, ctx=ast.Load()),
+            args=[ast.Constant("constant"), node],
+            keywords=[]
+        )
+
+
+
+def expr_checker(expr, get_attr, allow_function_calls=True, check_type=ast_default_check_type,
+                 check_function=ast_default_test):
+    code = f"""{dedent(getsource(check_type))}
```

Comment:
this won't work if you do `eval`

maybe add a parameter, if it's true, it return a dict containing the functions to be added in the execution context